brace-style: Having a catch method on a plain ES6 class throws `Closing curly brace does not appear on the same line as the subsequent brace style`

<!--
This template is for bug reports. If you are reporting a bug, please continue on. If you are here for another reason, please see below:

1. To propose a new rule: http://eslint.org/docs/developer-guide/contributing/new-rules
2. To request a change: http://eslint.org/docs/developer-guide/contributing/changes
3. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**Tell us about your environment**

* **ESLint Version: 3.14.0**
* **Node Version: 4.4.7**
* **npm Version: 2.15.8**

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint v5.0.4
**Please show your full configuration:**

<!-- Paste your configuration below: -->
```
{
  // http://eslint.org/docs/rules/
  "parser": "babel-eslint",
  "ecmaFeatures": {
    "binaryLiterals": false,                    // enable binary literals
    "blockBindings": false,                     // enable let and const (aka block bindings)
    "defaultParams": false,                     // enable default function parameters
    "forOf": false,                             // enable for-of loops
    "generators": false,                        // enable generators
    "modules": true,                            // enable ES Module System
    "objectLiteralComputedProperties": false,   // enable computed object literal property names
    "objectLiteralDuplicateProperties": false,  // enable duplicate object literal properties in strict mode
    "objectLiteralShorthandMethods": false,     // enable object literal shorthand methods
    "objectLiteralShorthandProperties": false,  // enable object literal shorthand properties
    "octalLiterals": false,                     // enable octal literals
    "regexUFlag": false,                        // enable the regular expression u flag
    "regexYFlag": false,                        // enable the regular expression y flag
    "templateStrings": false,                   // enable template strings
    "unicodeCodePointEscapes": false,           // enable code point escapes
    "jsx": false                                // enable JSX
  },

  "env": {
    "es6": true,         // all es6 features except modules
    "browser": true,     // browser global variables.
    "node": true,        // Node.js global variables and Node.js-specific rules.
    "jasmine": true,     // adds all of the Jasmine testing global variables for version 1.3 and 2.0.
    "phantomjs": true,   // phantomjs global variables.
    "jquery": true,      // jquery global variables.
  },

  "globals": {
    "document": true,
    "angular": true,
    "$": true
  },

  "plugins": [
    // e.g. "react" (must run `npm install eslint-plugin-react` first)
  ],

  "rules": {
    ////////// Possible Errors ////////

    // http://eslint.org/docs/rules/ //

    // 0 = 'off', 1 = 'warn', 2 = 'error' //
    "indent": [2, 2],
    "comma-dangle": 2,            // disallow trailing commas in object literals
    "no-cond-assign": 1,          // disallow assignment in conditional expressions
    "no-console": 1,              // disallow use of console (off by default in the node environment)
    "no-constant-condition": 2,   // disallow use of constant expressions in conditions
    "no-control-regex": 0,    // disallow control characters in regular expressions
    "no-debugger": 2,             // disallow use of debugger
    "no-dupe-keys": 2,            // disallow duplicate keys when creating object literals
    "no-empty": 2,                // disallow empty statements
    "no-empty-character-class": 2,// disallow the use of empty character classes in regular expressions
    "no-ex-assign": 2,            // disallow assigning to the exception in a catch block
    "no-extra-boolean-cast": 1 ,  // disallow double-negation boolean casts in a boolean context
    "no-extra-parens": 1,         // disallow unnecessary parentheses (off by default)
    "no-extra-semi": 2,           // disallow unnecessary semicolons
    "no-func-assign": 2,          // disallow overwriting functions written as function declarations
    "no-inner-declarations": 2,   // disallow function or variable declarations in nested blocks
    "no-invalid-regexp": 2,       // disallow invalid regular expression strings in the RegExp constructor
    "no-irregular-whitespace": 2, // disallow irregular whitespace outside of strings and comments
    "no-negated-in-lhs": 2,       // disallow negation of the left operand of an in expression
    "no-obj-calls": 2,            // disallow the use of object properties of the global object (Math and JSON) as functions
    "no-regex-spaces": 1,         // disallow multiple spaces in a regular expression literal
    "no-sparse-arrays": 2,        // disallow sparse arrays
    "no-unreachable": 2,          // disallow unreachable statements after a return, throw, continue, or break statement
    "use-isnan": 2,               // disallow comparisons with the value NaN
    "valid-jsdoc": 1,             // Ensure JSDoc comments are valid (off by default)
    "valid-typeof": 2,            // Ensure that the results of typeof are compared against a valid string


    ////////// Best Practices //////////

    "block-scoped-var": 2,      // treat var statements as if they were block scoped (off by default)
    "complexity": [2, {
      "max": 3
    }],                         // specify the maximum cyclomatic complexity allowed in a program (off by default)
    "consistent-return": 0,     // require return statements to either always or never specify values
    "curly": 2,                 // specify curly brace conventions for all control statements
    "default-case": 2,          // require default case in switch statements (off by default)
    "dot-notation": 2,          // encourages use of dot notation whenever possible
    "eqeqeq": 2,                // require the use of === and !==
    "guard-for-in": 1,          // make sure for-in loops have an if statement (off by default)
    "no-alert": 2,              // disallow the use of alert, confirm, and prompt
    "no-caller": 2,             // disallow use of arguments.caller or arguments.callee
    "no-div-regex": 0,          // disallow division operators explicitly at beginning of regular expression (off by default)
    "no-else-return":2,         // disallow else after a return in an if (off by default)
    "no-labels": 2,             // disallow use of labels for anything other then loops and switches
    "no-eq-null": 2,            // disallow comparisons to null without a type-checking operator (off by default)
    "no-eval": 2,               // disallow use of eval()
    "no-extend-native": 2,      // disallow adding to native types
    "no-extra-bind": 2,         // disallow unnecessary function binding
    "no-fallthrough": 2,        // disallow fallthrough of case statements
    "no-floating-decimal": 2,   // disallow the use of leading or trailing decimal points in numeric literals (off by default)
    "no-implied-eval": 2,       // disallow use of eval()-like methods
    "no-iterator": 2,           // disallow usage of __iterator__ property
    "no-lone-blocks": 2,        // disallow unnecessary nested blocks
    "no-loop-func": 2,          // disallow creation of functions within loops
    "no-multi-spaces": 0,       // disallow use of multiple spaces
    "no-multi-str": 2,          // disallow use of multiline strings
    "no-native-reassign": 2,    // disallow reassignments of native objects
    "no-new": 2,                // disallow use of new operator when not part of the assignment or comparison
    "no-new-func": 2,           // disallow use of new operator for Function object
    "no-new-wrappers": 2,       // disallows creating new instances of String, Number, and Boolean
    "no-octal": 2,              // disallow use of octal literals
    "no-octal-escape": 1,       // disallow use of octal escape sequences in string literals, such as var foo = "Copyright \251";
    "no-process-env": 1,        // disallow use of process.env (off by default)
    "no-proto": 1,              // disallow usage of __proto__ property
    "no-redeclare":2,           // disallow declaring the same variable more then once
    "no-return-assign": 2,      // disallow use of assignment in return statement
    "no-self-assign": 2,        // disallow assignment operators in return statements
    "no-script-url": 2,         // disallow use of javascript: urls.
    "no-self-compare": 2,       // disallow comparisons where both sides are exactly the same (off by default)
    "no-sequences": 2,          // disallow use of comma operator
    "no-unused-expressions": 2, // disallow usage of expressions in statement position
    "no-void": 0,               // disallow use of void operator (off by default)
    "no-warning-comments": 0,   // disallow usage of configurable warning terms in comments, e.g. TODO or FIXME (off by default)
    "no-with": 0,               // disallow use of the with statement
    "radix": 2,                 // require use of the second argument for parseInt() (off by default)
    "vars-on-top": 2,           // requires to declare all vars on top of their containing scope (off by default)
    "wrap-iife": 0,             // require immediate function invocation to be wrapped in parentheses (off by default)
    "yoda": 1,                  // require or disallow Yoda conditions


    ////////// Strict Mode //////////

    "strict": 2,          // controls location of Use Strict Directives

    ////////// Variables //////////
    "init-declarations": 0,           // require or disallow initialization in var declarations
    "no-catch-shadow": 2,             // disallow the catch clause parameter name being the same as a variable in the outer scope (off by default in the node environment)
    "no-delete-var": 2,               // disallow deletion of variables
    "no-label-var": 2,                // disallow labels that share a name with a variable
    "no-shadow": 2,                   // disallow declaration of variables already declared in the outer scope
    "no-shadow-restricted-names": 2,  // disallow shadowing of names such as arguments
    "no-undef": 2,                    // disallow use of undeclared variables unless mentioned in a /*global */ block
    "no-undef-init": 2,               // disallow use of undefined when initializing variables
    "no-undefined": 2,                // disallow use of undefined variable (off by default)
    "no-unused-vars": 2,              // disallow declaration of variables that are not used in the code
    "no-use-before-define": 2,        // disallow use of variables before they are defined


    ////////// Node.js //////////

    "handle-callback-err": 0,   // enforces error handling in callbacks (off by default) (on by default in the node environment)
    "no-mixed-requires": 0,     // disallow mixing regular variable and require declarations (off by default) (on by default in the node environment)
    "no-new-require": 0,        // disallow use of new operator with the require function (off by default) (on by default in the node environment)
    "no-path-concat": 0,        // disallow string concatenation with __dirname and __filename (off by default) (on by default in the node environment)
    "no-process-exit": 0,       // disallow process.exit() (on by default in the node environment)
    "no-restricted-modules": 0, // restrict usage of specified node modules (off by default)
    "no-sync": 0,               // disallow use of synchronous methods (off by default)


    ////////// Stylistic Issues //////////

    "brace-style": [2, "1tbs"],       // enforce one true brace style (off by default)
    "camelcase": [2, {
      properties: "always"
    }],                                     // require camel case names
    "comma-spacing": [2, {
      "before": false,
      "after": true
    }],                                     // enforce spacing before and after comma
    "comma-style": [2, "last"],       // enforce one true comma style (off by default)
    "consistent-this": [2, "_this"],  // enforces consistent naming when capturing the current execution context (off by default)
    "eol-last": 2,                          // enforce newline at the end of file, with no multiple empty lines
    "func-names": 0,                        // require function expressions to have a name (off by default)
    "func-style": 0,                        // enforces use of function declarations or expressions (off by default)
    "key-spacing": [2, {
      "beforeColon": false,
      "afterColon": true
    }],                                     // enforces spacing between keys and values in object literal properties
    "max-nested-callbacks": [2, 3],   // specify the maximum depth callbacks can be nested (off by default)
    "new-cap": 2,                           // require a capital letter for constructors
    "new-parens": 2,                        // disallow the omission of parentheses when invoking a constructor with no arguments
    "no-array-constructor": 2,              // disallow use of the Array constructor
    "no-inline-comments": 0,                // disallow comments inline after code (off by default)
    "no-lonely-if": 2,                      // disallow if as the only statement in an else block (off by default)
    "no-mixed-spaces-and-tabs": 2,          // disallow mixed spaces and tabs for indentation
    "no-multiple-empty-lines": 2,           // disallow multiple empty lines (off by default)
    "no-nested-ternary": 1,                 // disallow nested ternary expressions (off by default)
    "no-new-object": 2,                     // disallow use of the Object constructor
    "semi-spacing": [2, {
      "before": false,
      "after": true
    }],                                     // disallow space before semicolon
    "no-spaced-func": 2,                    // disallow space between function identifier and application
    "no-ternary": 0,                        // disallow the use of ternary operators (off by default)
    "no-trailing-spaces": 2,                // disallow trailing whitespace at the end of lines
    "no-underscore-dangle": 0,              // disallow dangling underscores in identifiers
    "one-var": 0,                           // allow just one var statement per function (off by default)
    "operator-assignment": 0,               // require assignment operator shorthand where possible or prohibit it entirely (off by default)
    "padded-blocks": 0,                     // enforce padding within blocks (off by default)
    "quote-props": 0,                       // require quotes around object literal property names (off by default)
    "quotes": [2, "single"],          // specify whether double or single quotes should be used
    "semi": [2, "always"],            // require or disallow use of semicolons instead of ASI
    "sort-vars": 2,                         // sort variables within the same declaration block (off by default)
    "space-before-function-paren": [
      2, "never"
    ],                                      // require a space after function names (off by default)
    "keyword-spacing": 2,                   // require a space after certain keywords (off by default)
    "space-before-blocks": 2,               // require or disallow space before blocks (off by default)
    "object-curly-spacing": [
      2, "always"
    ],                                      // require or disallow spaces around object literals
    "array-bracket-spacing": [
      2, "never"
    ],                                      // require or disallow spacing around array brackets
    "space-in-parens": [
      2, "never"
    ],                                      // require or disallow spaces inside parentheses (off by default)
    "space-infix-ops": 2,                   // require spaces around operators
    "space-unary-ops": [
      2, {
        "words": true,
        "nonwords": false
      }
    ],                                      // Require or disallow spaces before/after unary operators (words on by default, nonwords off by default)
    "spaced-comment": 2,                    // require or disallow a space immediately following the // in a line comment (off by default)
    "wrap-regex": 0,                        // require regex literals to be wrapped in parentheses (off by default)


    ////////// ECMAScript 6 //////////

    "no-var": 2,          // require let or const instead of var (off by default)
    "generator-star": 0,  // enforce the position of the * in generator functions (off by default)


    ////////// Legacy //////////

    "max-depth": 0,       // specify the maximum depth that blocks can be nested (off by default)
    "max-len": 0,         // specify the maximum length of a line in your program (off by default)
    "max-params": 0,      // limits the number of parameters that can be used in the function declaration. (off by default)
    "max-statements": 0,  // specify the maximum number of statement allowed in a function (off by default)
    "no-bitwise": 0,      // disallow use of bitwise operators (off by default)
    "no-plusplus": 0      // disallow use of unary operators, ++ and -- (off by default)
  }
}



```

**What did you do? Please include the actual source code causing the issue.**
We are mocking a promise interface within a backing object for our unit tests as part of a fake HTTP backend implementation.  It seems as though since v3.14.0 eslint thinks the catch method we have mocked here should always be a chained method call on a promise object.

<!-- Paste the source code below: -->
```js
export default class {

  constructor(data) {
    this.data = data;
  }

  then(callback) {
    this.data = callback(this.data);
    return this;
  }

  catch(callback) {
    callback({});
    return this;
  }
}
```

**What did you expect to happen?**
For eslint to recognise this as a plain ES6 class and lint it as such

**What actually happened? Please include the actual, raw output from ESLint.**

> Package.json name and version, and paths obfuscated here

```
[10:55:00] Using gulpfile ~/path/to/project/root/gulpfile.js
[10:55:00] Starting 'eslint'...
[10:55:02] ~/path/to/project/root/app/js/tests/backing/promises/fake-promise.js
  15:3  error  Closing curly brace does not appear on the same line as the subsequent block  brace-style

✖ 1 problem (1 error, 0 warnings)

[10:55:02] 'eslint' errored after 1.9 s
[10:55:02] ESLintError in plugin 'gulp-eslint'
Message:
    Failed with 1 error

npm ERR! Darwin 15.5.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "eslint"
npm ERR! node v4.4.7
npm ERR! npm  v2.15.8
npm ERR! code ELIFECYCLE
npm ERR! Project-name@version eslint: `gulp eslint`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the Project-name@version eslint script 'gulp eslint'.
npm ERR! This is most likely a problem with the Project-name package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     gulp eslint
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs Project-name
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!
npm ERR!     npm owner ls Project-name
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     ~/path/to/project/root/npm-debug.log
```

--- a/lib/rules/brace-style.js
+++ b/lib/rules/brace-style.js
@@ -78,7 +78,6 @@ module.exports = {
             const tokenBeforeOpeningCurly = sourceCode.getTokenBefore(openingCurly);
             const tokenAfterOpeningCurly = sourceCode.getTokenAfter(openingCurly);
             const tokenBeforeClosingCurly = sourceCode.getTokenBefore(closingCurly);
-            const tokenAfterClosingCurly = sourceCode.getTokenAfter(closingCurly);
             const singleLineException = params.allowSingleLine && astUtils.isTokenOnSameLine(openingCurly, closingCurly);
 
             if (style !== "allman" && !astUtils.isTokenOnSameLine(tokenBeforeOpeningCurly, openingCurly)) {
@@ -112,23 +111,30 @@ module.exports = {
                     fix: fixer => fixer.insertTextBefore(closingCurly, "\n")
                 });
             }
+        }
 
-            if (tokenAfterClosingCurly && tokenAfterClosingCurly.type === "Keyword" && new Set(["else", "catch", "finally"]).has(tokenAfterClosingCurly.value)) {
-                if (style === "1tbs" && !astUtils.isTokenOnSameLine(closingCurly, tokenAfterClosingCurly)) {
-                    context.report({
-                        node: closingCurly,
-                        message: CLOSE_MESSAGE,
-                        fix: removeNewlineBetween(closingCurly, tokenAfterClosingCurly)
-                    });
-                }
+        /**
+        * Validates the location of a token that appears before a keyword (e.g. a newline before `else`)
+        * @param {Token} curlyToken The closing curly token. This is assumed to precede a keyword token (such as `else` or `finally`).
+        * @returns {void}
+        */
+        function validateCurlyBeforeKeyword(curlyToken) {
+            const keywordToken = sourceCode.getTokenAfter(curlyToken);
 
-                if (style !== "1tbs" && astUtils.isTokenOnSameLine(closingCurly, tokenAfterClosingCurly)) {
-                    context.report({
-                        node: closingCurly,
-                        message: CLOSE_MESSAGE_STROUSTRUP_ALLMAN,
-                        fix: fixer => fixer.insertTextAfter(closingCurly, "\n")
-                    });
-                }
+            if (style === "1tbs" && !astUtils.isTokenOnSameLine(curlyToken, keywordToken)) {
+                context.report({
+                    node: curlyToken,
+                    message: CLOSE_MESSAGE,
+                    fix: removeNewlineBetween(curlyToken, keywordToken)
+                });
+            }
+
+            if (style !== "1tbs" && astUtils.isTokenOnSameLine(curlyToken, keywordToken)) {
+                context.report({
+                    node: curlyToken,
+                    message: CLOSE_MESSAGE_STROUSTRUP_ALLMAN,
+                    fix: fixer => fixer.insertTextAfter(curlyToken, "\n")
+                });
             }
         }
 
@@ -154,6 +160,24 @@ module.exports = {
                 const openingCurly = sourceCode.getTokenBefore(node.cases.length ? node.cases[0] : closingCurly);
 
                 validateCurlyPair(openingCurly, closingCurly);
+            },
+            IfStatement(node) {
+                if (node.consequent.type === "BlockStatement" && node.alternate) {
+
+                    // Handle the keyword after the `if` block (before `else`)
+                    validateCurlyBeforeKeyword(sourceCode.getLastToken(node.consequent));
+                }
+            },
+            TryStatement(node) {
+
+                // Handle the keyword after the `try` block (before `catch` or `finally`)
+                validateCurlyBeforeKeyword(sourceCode.getLastToken(node.block));
+
+                if (node.handler && node.finalizer) {
+
+                    // Handle the keyword after the `catch` block (before `finally`)
+                    validateCurlyBeforeKeyword(sourceCode.getLastToken(node.handler.body));
+                }
             }
         };
     }

--- a/tests/lib/rules/brace-style.js
+++ b/tests/lib/rules/brace-style.js
@@ -170,6 +170,30 @@ ruleTester.run("brace-style", rule, {
                 {
                 }
             }
+        `,
+
+        // https://github.com/eslint/eslint/issues/7974
+        `
+          class Ball {
+            throw() {}
+            catch() {}
+          }
+        `,
+        `
+          ({
+            and() {},
+            finally() {}
+          })
+        `,
+        `
+          (class {
+            or() {}
+            else() {}
+          })
+        `,
+        `
+          if (foo) bar = function() {}
+          else baz()
         `
     ],
 

